instructions for udfs (#81)
confluentinc/tutorials: vscode-udfs/README.md

<!-- title: How to use Confluent for VS Code to create a sample Java UDF in Confluent Cloud for Apache Flink -->
<!-- description: In this tutorial, learn how to develop, upload, register and execute Flink UDFs with Confluent for VS Code, with step-by-step instructions. -->

# How to use Confluent for VS Code to create a sample Java UDF in Confluent Cloud for Apache Flink

In this tutorial, you'll follow along step-by-step to build a Java UDF, then deploy it in Confluent Cloud for Apache Flink. 

> ## Pre-requisites

Reviewer: ```suggestion
## Prerequisites
```

1. You will need a Confluent Cloud [account](https://www.confluent.io/confluent-cloud/tryfree/) if you haven't got one already. 

2. [Install Confluent for VS Code](https://docs.confluent.io/cloud/current/client-apps/vs-code-extension.html).

> 3. Get [Java version 11 or 17](https://www.oracle.com/java/technologies/).

Reviewer: ```suggestion
3. Get [Java version 11 or 17](https://openjdk.org/install/).
```

```

In the sidebar, click "Support" then "Generate Project From Template". Search "UDFs" and click on the UDF template generator. 

Navigate to the folder you generated the template in. 

This should produce a jar for each of the modules, which can be uploaded to Confluent Cloud.

For example:

```shell
> find . -wholename '*/target/*jar' | grep -v original
./udfs-simple/target/udfs-simple-1.0.0.jar
```

You should then upload the jars to Confluent Cloud, and create a UDF in your catalog.
More details on how to do that can be found
[here](https://docs.confluent.io/cloud/current/flink/how-to-guides/create-udf.html).

Here's an example of the command you'd run to register the function in your Flink workspace once you've uploaded the artifact following the instructions above:

```sql
CREATE FUNCTION sum_integers
AS 'io.confluent.udf.examples.log.LogSumScalarFunction'
USING JAR 'confluent-artifact://cfa-v7r61n';
```
Remember to replace `cfa-v7r61n` with your own artifact id.

Then you can test the function by running this statement:

```sql
SELECT sum_integers(CAST(5 AS INT), CAST(3 AS INT));
```

## Inspect the Code

In the `pom.xml` you'll see the following:

```xml

>     <dependencies>
>         <dependency>

Reviewer: ```suggestion
        <dependency>
```

            <groupId>org.apache.flink</groupId>
            <artifactId>flink-table-api-java</artifactId>
            <scope>provided</scope>
        </dependency>
```

This is vital to the execution of user-defined functions with Flink. This is where you'll get the base classes 
you need to implement a UDF. If you look at `LogSumScalarFunction`, you'll see an import of the `ScalarFunction` class:

```java
import org.apache.flink.table.functions.ScalarFunction;
```

> Then this class is extended to write the custom logging function below:

Reviewer: ```suggestion
Then this class is extended to sum two numbers and log messages:
```

        Date now = new Date();

        // You can choose the logging level for log messages.
        LOGGER.info(value + " info log messages by log4j logger --- " + now);
        LOGGER.error(value + " error log messages by log4j logger --- " + now);
        LOGGER.warn(value + " warn log messages by log4j logger --- " + now);
        LOGGER.debug(value + " debug log messages by log4j logger --- " + now);
        return a + b;
    }
}
```

In `Repeater`, you'll see the following:

```java
import org.apache.flink.table.annotation.DataTypeHint;
import org.apache.flink.table.annotation.FunctionHint;
import org.apache.flink.table.functions.TableFunction;
```

`Repeater` imports both annotations and functions in order to implement the table functions. In a similar pattern to scalar functions, the `TableFunction`
class is extended in order to write the custom function below.

```java
@FunctionHint(output = @DataTypeHint("ROW<val INT>"))
public class Repeater extends TableFunction<Row> {

    public void eval(Integer number, Integer times) {
        for (int i = 0; i < times; i++) {
            // use collect(...) to emit a row
            collect(Row.of(number));
        }
    }
}
```

> 

Reviewer: why include `Repeater`? It feels like there is missing context. I think remove this block since the run instructions only cover the scalar UDF

> <!-- title: How to create UDFs with Confluent for VS Code -->

Reviewer: I think that the title should mention CC for AF, and also make it clear that it's the sample template. i.e., this isn't explaining to users how to develop a UDF though arguably it should get into that more (comment on that below). I would suggest `How to use Confluent for VS Code to create a sample UDF in Confluent Cloud for Apache Flink` or `How to create a sample UDF in Confluent Cloud for Apache Flink using Confluent for VS Code`

<!-- description: In this tutorial, learn how to develop, upload, register and execute Flink UDFs with Confluent for VS Code, with step-by-step instructions. -->

> # VSCode Flink UDFs

Reviewer: change this title to match what you go with above. `How to use Confluent for VS Code to create a sample UDF in Confluent Cloud for Apache Flink` is long but we're putting together 2 product names so I think it's ok. we want to avoid people thinking that the tutorial is for OSS AF so it should call these things out

> [Install Confluent for VS Code](https://docs.confluent.io/cloud/current/client-apps/vs-code-extension.html).

Reviewer: Add a prereqs section that includes:
* Confluent Cloud account
* installing the extension
* Java version 11 or 17 (I got bit by this, happened to have java 8 enabled and it wont compile with 8)
* Maven

> In the VS Code Activity Bar, click the Confluent icon.

Reviewer: the rest of this can be in a separate section `Build and Deploy UDF` or something. So overall, the format I would suggest is:

```
# How to use Confluent for VS Code to create a sample UDF in Confluent Cloud for Apache Flink

In this tutorial, ...

## Prerequisites

...

## Build and Deploy UDF

...
```

![Activity Bar Selection](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-produce-and-consume/img/activity-bar-selection.png)

If you have many extensions installed, you may need to click … to access 'Additional Views' and select 'Confluent' from the context menu.

![Activity Bar Selection via Additional Views](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-produce-and-consume/img/via-views.png)

In the Side Bar, click 'Connect to Confluent Cloud', and in the permission dialog, click 'Allow'.

![Connect to Confluent Cloud](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-produce-and-consume/img/connect-to-cc.png)

> ![Allow opening in new window](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-ccloud-quickstart/img/new-window.png)

Reviewer: ```suggestion
![Allow opening in new window](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-produce-and-consume/img/new-window.png)
```

The web browser opens to the Confluent Cloud login page.

Enter your credentials and click 'Log in'.

![Confluent Cloud Log In Page](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-produce-and-consume/img/login.png)

The web browser shows the 'Authentication Complete' page.

![Head to Confluent Cloud page](https://raw.githubusercontent.com/confluentinc/tutorials/master/vscode-produce-and-consume/img/auth-complete-page.png)

Open the Confluent Cloud Console and [follow these steps](https://docs.confluent.io/cloud/current/get-started/index.html#section-1-create-a-cluster-and-add-a-topic) to create an environment.

Next, create a Flink compute pool by following [these steps](https://docs.confluent.io/cloud/current/flink/operate-and-deploy/create-compute-pool.html).

While that pool is provisioning, return to your editor and confirm that your Confluent Cloud resources are displayed in the Side Bar.

Next, click on "Support" in the sidebar and then "Generate Project From Template". Search "UDFs" and select the Flink UDF template and select your desired directory. 

Navigate to the directory you generate the template in. Time to build the examples!

> ## Building the examples

Reviewer: this tutorial doesn't do much guide me to understand "how to develop" a UDF even though it says that above. I think that it should have a section called `Inspect the code` or something that shows users (with code snippets) the important things they need to know:

1. dependencies, i.e. `org.apache.flink:flink-table-api-java` which provides the base classes that you extend to implement a UDF
2. the interface, i.e., "you have to extend ScalarFunction and implement a public method named `eval` etc etc and here is what that looks like for one of the example UDFs that simply sums two integers."

First, generate the maven wrapper jar:

```shell
mvn wrapper:wrapper 
```

You may need to run:

```shell
chmod +x mvnw
```

then: 

```shell
./mvnw clean package
```

In the sidebar, click "Support" then "Generate Project From Template". Search "UDFs" and click on the UDF template generator. 

Navigate to the folder you generated the template in. 

This should produce a jar for each of the modules, which can be uploaded to Confluent Cloud.

For example:

```shell
> find . -wholename '*/target/*jar' | grep -v original
./udfs-simple/target/udfs-simple-1.0.0.jar
./udfs-with-dependencies/target/udfs-with-dependencies-1.0.0.jar
```

You should then upload the jars to Confluent Cloud, and create a UDF in your catalog.
More details on how to do that can be found

> [here](https://docs.confluent.io/cloud/current/flink/how-to-guides/create-udf.html).

Reviewer: This tutorial really only needs `./udfs-simple/target/udfs-simple-1.0.0.jar` so I think the instructions in this block should focus down to that one jar.

a benefit of this is that it will skirt an ambiguity problem below, which is that if I upload both, I don't know which artifact ID (`cfa-abcdef`) to use  when creating the UDF. only the simple jar will work. If I try to use the `udfs-with-dependencies-1.0.0.jar` I hit an error about the class not being found

Here's an example of the command you'd run to register the function in your Flink workspace once you've uploaded the artifact following the instructions above:

```sql

> CREATE FUNCTION my_udf

Reviewer: give a descriptive function name to leave out guesswork and also promote good naming practices, e.g.:
```suggestion
CREATE FUNCTION sum_integers
```

AS 'io.confluent.udf.examples.log.LogSumScalarFunction'
USING JAR 'confluent-artifact://cfa-v7r61n';
```
Remember to replace `cfa-v7r61n` with your own artifact id.

Then you can test the function by running this statement:

```sql

> SELECT my_udf(CAST(5 AS INT), CAST(3 AS INT));

Reviewer: (or whatever name you go with)
```suggestion
SELECT sum_integers(CAST(5 AS INT), CAST(3 AS INT));
```
